File: Demo/Sources/ViewControllers/Choise/ChoiseViewController.m
//
//  Globus
//

#import "ChoiseViewController.h"
#import "ChoiseCellTableViewCell.h"
#import "ChoiseViewModel.h"

#import "MainViewController.h"
#import "LabelViewController.h"
#import "ButtonViewController.h"
#import "ButtonImageViewController.h"
#import "ButtonBadgeViewController.h"
#import "ImageViewController.h"
#import "ImagePickerCropViewController.h"
#import "TextFieldViewController.h"
#import "DateFieldViewController.h"
#import "ListFieldViewController.h"
#import "PhoneFieldViewController.h"
#import "TextViewController.h"
#import "SpinnerViewController.h"
#import "ActivityViewController.h"
#import "ScrollViewController.h"
#import "TransitionRootViewController.h"
#import "PageControlViewController.h"
#import "NotificationsViewController.h"
#import "DialogRootViewController.h"
#import "ListDataViewController.h"
#import "SimpleDataViewController.h"
#import "CalendarDataViewController.h"

@interface ChoiseViewController () < UITableViewDataSource, UITableViewDelegate > {
    NSArray< ChoiseViewModel* >* _data;
}

@property(nonatomic, weak) IBOutlet UITableView* tableView;

@end

@implementation ChoiseViewController

#pragma mark - UIViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _data = @[
        [ChoiseViewModel viewModelWithTitle:@"Main" viewController:MainViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Label" viewController:LabelViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Button" viewController:ButtonViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Button & Image" viewController:ButtonImageViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Button & Badge" viewController:ButtonBadgeViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Image" viewController:ImageViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Image picker & crop" viewController:ImagePickerCropViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Text field" viewController:TextFieldViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Date field" viewController:DateFieldViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"List field" viewController:ListFieldViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Phone field" viewController:PhoneFieldViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Text view" viewController:TextViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Spinner view" viewController:SpinnerViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Activity view" viewController:ActivityViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"ScrollView" viewController:ScrollViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Transition" viewController:TransitionRootViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"PageControl" viewController:PageControlViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Notifications" viewController:NotificationsViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"Dialog" viewController:DialogRootViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"ListDataView" viewController:ListDataViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"SimpleDataView" viewController:SimpleDataViewController.class],
        [ChoiseViewModel viewModelWithTitle:@"CalendarDataView" viewController:CalendarDataViewController.class],
    ];
    
    [_tableView glb_registerCell:ChoiseCellTableViewCell.class];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView*)tableView numberOfRowsInSection:(NSInteger)section {
    return (NSInteger)(_data.count);
}

- (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath {
    return [tableView glb_dequeueReusableCell:ChoiseCellTableViewCell.class];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView*)tableView willDisplayCell:(UITableViewCell*)cell forRowAtIndexPath:(NSIndexPath*)indexPath {
    NSUInteger index = (NSUInteger)(indexPath.row);
    if([cell isKindOfClass:ChoiseCellTableViewCell.class] == YES) {
        ChoiseCellTableViewCell* choiseCell = (ChoiseCellTableViewCell*)cell;
        [choiseCell configureWithChoiseViewModel:_data[index]];
    }
}

- (void)tableView:(UITableView*)tableView didSelectRowAtIndexPath:(NSIndexPath*)indexPath {
    GLBSlideViewController* svc = self.glb_slideViewController;
    if([svc.centerViewController isKindOfClass:GLBNavigationViewController.class] == YES) {
        GLBNavigationViewController* nvc = svc.centerViewController;
        
        NSUInteger index = (NSUInteger)(indexPath.row);
        GLBViewController* vc = [_data[index] instantiateViewController];
        if(vc != nil) {
            [nvc setViewControllers:@[ vc ] animated:YES];
        }
        [svc hideLeftViewControllerAnimated:YES complete:nil];
    }
}

@end

File: Demo/Sources/ViewControllers/Choise/ChoiseViewController.swift
//
//  Globus
//

import Globus

class ChoiseViewController: GLBViewController {
    
    // MARK: - Outlet property
    
    @IBOutlet fileprivate weak var tableView: UITableView!
    
    // MARK: - Private property
    
    var data: Array< ChoiseViewModel > = [
        ChoiseControllerViewModel< MainViewController >(title:"Main"),
        ChoiseControllerViewModel< LabelViewController >(title:"Label"),
        ChoiseControllerViewModel< ButtonViewController >(title:"Button"),
        ChoiseControllerViewModel< ButtonImageViewController >(title:"Button & Image"),
        ChoiseControllerViewModel< ButtonBadgeViewController >(title:"Button & Badge"),
        ChoiseControllerViewModel< ImageViewController >(title:"Image"),
        ChoiseControllerViewModel< ImagePickerCropViewController >(title:"Image picker & crop"),
        ChoiseControllerViewModel< TextFieldViewController >(title:"Text field"),
        ChoiseControllerViewModel< DateFieldViewController >(title:"Date field"),
        ChoiseControllerViewModel< ListFieldViewController >(title:"List field"),
        ChoiseControllerViewModel< PhoneFieldViewController >(title:"Phone field"),
        ChoiseControllerViewModel< TextViewController >(title:"Text view"),
        ChoiseControllerViewModel< SpinnerViewController >(title:"Spinner view"),
        ChoiseControllerViewModel< ActivityViewController >(title:"Activity view"),
        ChoiseControllerViewModel< ScrollViewController >(title:"ScrollView"),
        ChoiseControllerViewModel< TransitionRootViewController >(title:"Transition"),
        ChoiseControllerViewModel< PageControlViewController >(title:"PageControl"),
        ChoiseControllerViewModel< NotificationsViewController >(title:"Notifications"),
        ChoiseControllerViewModel< DialogRootViewController >(title:"Dialog"),
        ChoiseControllerViewModel< ListDataViewController >(title:"ListDataView"),
        ChoiseControllerViewModel< SimpleDataViewController >(title:"SimpleDataView"),
        ChoiseControllerViewModel< CalendarDataViewController >(title:"CalendarDataView"),
    ]
    
    //MARK: - UIViewController
    
    public override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.glb_registerCell(ChoiseCellTableViewCell.self)
    }
    
    // MARK: - GLBNibExtension
    
    public override static func nibName() -> String {
        return "ChoiseViewController-Swift"
    }
    
}

//MARK: - UITableViewDataSource

extension ChoiseViewController: UITableViewDataSource {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.data.count;
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return tableView.glb_dequeueReusableCell(ChoiseCellTableViewCell.self)!
    }
    
}

//MARK: - UITableViewDelegate

extension ChoiseViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let svc = self.glb_slideViewController {
            if let nvc = svc.centerViewController as? UINavigationController {
                let viewModel = self.data[indexPath.row]
                if let vc = viewModel.instantiateViewController() {
                    nvc.setViewControllers([ vc ], animated: true)
                }
            }
            svc.hideLeftViewController(animated: true, complete: nil)
        }
    }
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if let choiseCell = cell as? ChoiseCellTableViewCell {
            let viewModel = self.data[indexPath.row]
            choiseCell.configure(viewModel: viewModel)
        }
    }
    
}
